packages/monaco/src/browser/monaco-editor-model.ts

```diff
@@ -33,29 +35,79 @@ export class MonacoEditorModel implements ITextEditorModel {
     protected readonly onDidSaveModelEmitter = new Emitter<monaco.editor.IModel>();
     protected readonly onWillSaveModelEmitter = new Emitter<WillSaveModelEvent>();
 
-    constructor(protected readonly resource: Resource) {
+    constructor(
+        protected readonly resource: Resource,
+        protected readonly m2p: MonacoToProtocolConverter,
+        protected readonly p2m: ProtocolToMonacoConverter
+    ) {
         this.toDispose.push(resource);
         this.toDispose.push(this.toDisposeOnAutoSave);
         this.toDispose.push(this.onDidSaveModelEmitter);
         this.toDispose.push(this.onWillSaveModelEmitter);
         this.resolveModel = resource.readContents().then(content => this.initialize(content));
     }
 
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
     /**
      * #### Important
      * Only this method can create an instance of `monaco.editor.IModel`,
      * there should not be other calls to `monaco.editor.createModel`.
      */
-    protected initialize(content: string) {
+    protected initialize(content: string): void {
         if (!this.toDispose.disposed) {
             this.model = monaco.editor.createModel(content, undefined, monaco.Uri.parse(this.resource.uri.toString()))
             this.toDispose.push(this.model);
-            this.toDispose.push(this.model.onDidChangeContent(event => this.doAutoSave()));
+            this.toDispose.push(this.model.onDidChangeContent(event => this.markAsDirty()));
+            if (this.resource.onDidChangeContents) {
+                this.toDispose.push(this.resource.onDidChangeContents(() => this.sync()));
+            }
         }
     }
 
-    dispose(): void {
-        this.toDispose.dispose();
+    protected _dirty = false;
+    get dirty(): boolean {
+        return this._dirty;
+    }
+    protected setDirty(dirty: boolean): void {
+        this._dirty = dirty;
+        this.onDirtyChangedEmitter.fire(undefined);
+    }
+
+    protected readonly onDirtyChangedEmitter = new Emitter<void>();
```

Comment:
Please pull these up if you do not have a good reason to keep them there.: `onDirtyChangedEmitter`, `_dirty` and `synchronizing`. Thanks!

Reply:
They belong to the part where they used. Why is it important to declare everything in one place?